resolved issue 8884 - the properties don't return all properties and imp...

montegoulding opened the pull request:
...lemented a few changes to the array.

Additionally changed from colours and patterns list to individual elements so it's easier to work with the array. Changed to return unicode variants where they double up so no data is lost in native formats. Added resizeQuality, paintCompression, stack filename, externals. The array is now multi-dimensional where properties are themselves arrays. Finally, added the optional 'effective' adjective to the properties.

Sending the pull request against release-6.0.1 for simplicity but don't really expect it to be looked at until after release as there could be some (albeit minor considering the low level of use of the property) backwards compatibility concerns in this one. I think most users of `the properties` would be quite keen on the changes.

Commit:
resolved issue 8884 - the properties don't return all properties and implemented a few changes to the array.

Additionally changed from colours and patterns list to individual elements so it's easier to work with the array. Changed to return unicode variants where they double up so no data is lost in native formats. Added resizeQuality, paintCompression, stack filename, externals. The array is now multi-dimensional where properties are themselves arrays. Finally, added the optional 'effective' adjective to the properties.

runrevmark commented:
Since you were looking to do some more work on this, I'll leave it for now... Seems like a good candidate for the development branch (and thus next non-maintenance release) when its ready :)

Commit:
added unbounded scroll properties to group

Commit:
a few more tweaks

Commit:
made some more changes to the properties array and setting properties

getting properties added fullscreen, editMenus, shadowOffset, systemWindow, layerMode, padding. removed some read only properties paintCompression, size

Changed properties array setter to set rect and style first to ensure they are set before other properties that might change as a result of them.. for example, setting the rect after the gradient changes the gradient. The style seems to change other properties like the textAlign of buttons which possibly needs to be investigated.

Commit:
removed read only stack vScroll, image frameCount. scrollbar orientation. removed session only button visited

Commit:
setting textSize first before iterating the array to ensure textHeight isn't overridden

Commit:
set width and height first to ensure they don't break gradients and to ensure left, top etc are applied correctly

Commit:
a little bit of refactoring on set props and move to returning unicode props only if hasunicode(). Also removed card rect as read only.

Commit:
more enhancements for unicode properties which are now individually converted to native if possible using MCExecPoint::trytoconvertutf16tonative

runrevmark commented on engine/src/props.cpp:
```diff
@@ -570,8 +767,45 @@ Exec_stat MCObject::getproparray(MCExecPoint &ep, uint4 parid)
 	{
-		getprop(parid, (Properties)table[tablesize].value, ep, False);
-		v->store_element(ep, table[tablesize].token);
+        char * t_token = strdup(table[tablesize].token);
```
I don't believe there's any need to copy the string here - store_element() doesn't take ownership. You can just use the const strings. (Same is true with all assignements to t_token).

montegoulding replied:
ok, fixed

runrevmark commented on engine/src/variablearray.cpp:
```diff
@@ -1374,11 +1374,51 @@ Exec_stat MCVariableArray::setprops(uint4 parid, MCObject *optr)
+    {
+        e -> value . fetch(ep);
+        optr->setprop(parid, P_RECTANGLE, ep, False);
+        removehash(e);
```
This isn't a good idea as you are removing entries from the array. In many cases this will be referencing the array that is pointed to by an actual variable so you are deleting the keys from there too. e.g.

```
get the properties of me
set the properties of me to it
put "rect" is among the keys of it
```

Will result in false.

montegoulding replied:
oh, ok, I'll need to iterate through the array of pre-parse setprops then to make sure they aren't being set again..

runrevmark commented on engine/src/variablearray.cpp:
```diff
@@ -1374,11 +1374,51 @@ Exec_stat MCVariableArray::setprops(uint4 parid, MCObject *optr)
 	MCerrorlock++;
+    MCHashentry *e;
+    // set rect first so it doesn't break gradients
+    e = lookuphash(MCString("rect"),true,false);
```
There should be no need to use an MCString constructor here as one taking 'const char *' is a default one.

montegoulding replied:
ok

runrevmark commented on engine/src/variablearray.cpp:
```diff
@@ -1374,11 +1374,51 @@ Exec_stat MCVariableArray::setprops(uint4 parid, MCObject *optr)
 	MCExecPoint ep(optr, NULL, NULL);
 	MCRedrawLockScreen();
 	MCerrorlock++;
+    MCHashentry *e;
```
This sequence of lookups and then processing is all the same except for property and token - it would be better to use an array which is then iterated through in order. e.g.

```
static struct { Properties prop; const char *tag; } s_preprocess_props[] =
{
    { P_RECTANGLE, "rectangle" },
    ...
}
```

You can check the length of the array by using sizeof(s_preprocess_props) / sizeof(s_preprocess_props[0]).

montegoulding replied:
ah... yeah, I started with just one pre-parse set prop and ended up with a few but didn't refactor ;-)

montegoulding replied:
Hmm... should I only pre-process keys that are actually in `the properties` when you get them or should I handle cases where people have used a synonym. For example, you have used `rectangle` while `rect` is the key in `the properties`... I guess it's just another line in the array but for each array entry it's also a string comparison on every key in the array because I can't delete the key...

Commit:
a bit of refactoring and bugfixing as per comments on the pull request

runrevmark commented:
Hi Monte,

Sorry for taking a while to get back to looking at this - last week was a bit busy after getting back from holiday!

Anyway, I've taken another look and made a couple of comments on the code... It looks almost ready to go :)

Thanks,

Mark.

Commit:
change from uint1 to uindex_t and changed pre-processed handler to remove string comparisons

runrevmark commented:
Hi Monte

I've just pulled this in as branch 'feature-revised_properties_prop' and merged that into 'develop'.

Thanks!

Mark.
